sklearn/datasets/_base.py

```diff
@@ -1454,28 +1464,161 @@ def _fetch_remote(remote, dirname=None, n_retries=3, delay=1):
+        prefix=remote.filename + ".part_", dir=folder_path, delete=False
+    )
+    temp_file.close()
+    try:
+        temp_file_path = Path(temp_file.name)
+        while True:
+            try:
+                urlretrieve(remote.url, temp_file_path)
+                break
+            except (URLError, TimeoutError):
+                if n_retries == 0:
+                    # If no more retries are left, re-raise the caught exception.
+                    raise
+                warnings.warn(f"Retry downloading from url: {remote.url}")
+                n_retries -= 1
+                time.sleep(delay)
+
+        checksum = _sha256(temp_file_path)
+        if remote.checksum is not None and remote.checksum != checksum:
+            raise OSError(
+                f"The SHA256 checksum of {remote.filename} ({checksum}) "
+                f"differs from expected ({remote.checksum})."
+            )
+    except BaseException:
+        os.unlink(temp_file.name)
+        raise
+
+    # The following renaming is atomic whenever temp_file_path and
+    # file_path are on the same filesystem. This should be the case most of
+    # the time, but we still use shutil.move instead of os.rename in case
+    # they are not.
+    shutil.move(temp_file_path, file_path)
 
-    file_path = remote.filename if dirname is None else join(dirname, remote.filename)
-    while True:
-        try:
-            urlretrieve(remote.url, file_path)
-            break
-        except (URLError, TimeoutError):
-            if n_retries == 0:
-                # If no more retries are left, re-raise the caught exception.
-                raise
-            warnings.warn(f"Retry downloading from url: {remote.url}")
-            n_retries -= 1
-            time.sleep(delay)
-
-    checksum = _sha256(file_path)
-    if remote.checksum != checksum:
-        raise OSError(
-            "{} has an SHA256 checksum ({}) "
-            "differing from expected ({}), "
-            "file may be corrupted.".format(file_path, checksum, remote.checksum)
-        )
     return file_path
+
+
+def _slugify(value):
+    """Derive a name that is safe to use as filename from the given string.
+
+    Adapted from
+    https://github.com/django/django/blob/master/django/utils/text.py
+
+    Convert to ASCII if 'allow_unicode' is False. Convert spaces or repeated
+    dashes to single dashes. Remove characters that aren't alphanumerics,
+    underscores, or hyphens. Convert to lowercase. Also strip leading and
+    trailing whitespace, dashes, and underscores.
+
+    Note: this version keeps "." characters unchanged contrary to the django
+    version and replace other un-authorized characters by "_".
+    """
+    value = (
+        unicodedata.normalize("NFKD", value).encode("ascii", "ignore").decode("ascii")
+    )
+    value = re.sub(r"[^.\w\s-]", "_", value.lower())
+    value = re.sub(r"_+", "_", value)
+    return re.sub(r"[-\s]+", "-", value).strip("-_")
```

Comment:
this really isn't safe:

```py
In [2]: _slugify("نمونه.txt")
Out[2]: '.txt'
```

Reply:
ideally we should get something which is supported by users's FS, which is unicode in most cases these days I guess, and if not, replace the characters with something >128 ascii to indicate unsupported character replacement.

Reply:
> this really isn't safe:

I would argue that it's safe in the sense that this extreme normalization will not lead to an attempt to write a file in a different folder or with an invalid name. But I agree it can lead to unwanted collisions.

Actually, your example makes be realize that we need to strip leading "." (and probably trailing "." as well for the sake of consistency).

> ideally we should get something which is supported by users's FS, which is unicode in most cases these days

I think ascii conversion is useful when generating public URL components (a.k.a. slugs) from user submitted strings to avoid homograph attacks to serve malware. However it won't be a problem on a filesystem. Furthermore, modern OS should have no problem with arbitrary unicode names as long as punctuation is filtered out: 

https://stackoverflow.com/a/35352640/163740

I will update the function and the tests.